Need to refactor the test_bgp_speaker.py to take advantage of enhanced fixtures

<!--
If you are reporting a new issue, make sure that we do not have any duplicates
already open. You can ensure this by searching the issue list for this
repository. If there is a duplicate, please close your issue and add a comment
to the existing issue instead.

If you suspect your issue is a bug, please edit your issue description to
include the BUG REPORT INFORMATION shown below. If you fail to provide this
information within 7 days, we cannot debug your issue and will close it. We
will, however, reopen it if you later provide the information.

For more information about reporting issues, see
https://github.com/Azure/SONiC/wiki#report-issues

---------------------------------------------------
GENERAL SUPPORT INFORMATION
---------------------------------------------------

The GitHub issue tracker is for bug reports and feature requests.
General support can be found at the following locations:

- SONiC Support Forums - https://groups.google.com/forum/#!forum/sonicproject

---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------
Use the commands below to provide key information from your environment:
You do NOT have to include this information if this is a FEATURE REQUEST
-->

**Description**
The BGP Speaker testing has been refactored to use the pytest script. In case we want to re-use the existing testbed.csv, the BGP speaker case would fail with creating PTF Ansible host because the existing testbeds.csv does not have the "ptf" column.
The testbed_devices fixture in sonic-mgmt/tests/conftest.py has been enhanced to look for PTF information from the inventory file in case testbeds.csv does not have the ptf column.
The test_bgp_speaker.py script can be enhanced to take advantange of the testbed_device fixture.
Otherwise, it may fail with below error:

```
07:45:19 TASK [test : run py.test] ******************************************************
07:45:19 task path: /builds2/jenkins2/workspace/SONiC_Test-BGP-Speaker/ansible/roles/test/tasks/pytest_runner.yml:54
07:45:19 Sunday 24 November 2019  23:45:19 +0000 (0:00:00.170)       0:00:13.397 ******* 
07:45:19 ESTABLISH LOCAL CONNECTION FOR USER: root
07:45:19 r-tigris-04 EXEC ( umask 22 && mkdir -p "$( echo $HOME/.ansible/tmp/ansible-tmp-1574639119.47-13449679608551 )" && echo "$( echo $HOME/.ansible/tmp/ansible-tmp-1574639119.47-13449679608551 )" )
07:45:19 r-tigris-04 PUT /tmp/tmpLtHHhN TO /root/.ansible/tmp/ansible-tmp-1574639119.47-13449679608551/command
07:45:19 r-tigris-04 EXEC LANG=C LC_ALL=C ANSIBLE_LIBRARY=../ansible/library/ LC_MESSAGES=C /usr/bin/python /root/.ansible/tmp/ansible-tmp-1574639119.47-13449679608551/command; rm -rf "/root/.ansible/tmp/ansible-tmp-1574639119.47-13449679608551/" > /dev/null 2>&1
07:45:23 fatal: [r-tigris-04-t0]: FAILED! => {"changed": true, "cmd": "py.test test_bgp_speaker.py --testbed=r-tigris-04-t0 --inventory=../ansible/inventory --testbed_file=../ansible/testbed.csv --host-pattern=r-tigris-04-t0 -v", "delta": "0:00:04.192530", "end": "2019-11-24 23:45:23.718437", "failed": true, "invocation": {"module_args": {"_raw_params": "py.test test_bgp_speaker.py --testbed=r-tigris-04-t0 --inventory=../ansible/inventory --testbed_file=../ansible/testbed.csv --host-pattern=r-tigris-04-t0 -v", "_uses_shell": true, "chdir": "../tests/", "creates": null, "executable": null, "removes": null, "warn": true}, "module_name": "command"}, "rc": 1, "start": "2019-11-24 23:45:19.525907", "stderr": "", "stdout": "============================= test session starts ==============================\nplatform linux2 -- Python 2.7.12, pytest-4.6.5, py-1.8.0, pluggy-0.12.0 -- /usr/bin/python\ncachedir: .pytest_cache\nansible: 2.0.0.2\nrootdir: /builds2/jenkins2/workspace/SONiC_Test-BGP-Speaker/tests, inifile: pytest.ini\nplugins: ansible-2.0.2\ncollecting ... collected 1 item\n\ntest_bgp_speaker.py::test_bgp_speaker[True-False-1514] FAILED            [100%]\n\n=================================== FAILURES ===================================\n______________________ test_bgp_speaker[True-False-1514] _______________________\n\nlocalhost = <pytest_ansible.module_dispatcher.v2.ModuleDispatcherV2 object at 0x7f779dbfae10>\nansible_adhoc = <function init_host_mgr at 0x7f779dc055f0>\ntestbed = {'comment': 'Mellanox MTR testbed', 'dut': 'r-tigris-04', 'group-name': 'vm-t1', 'ptf_image_name': 'docker-ptf-mlnx', ...}\nipv4 = True, ipv6 = False, mtu = 1514\n\n    @pytest.mark.parametrize(\n        \"ipv4, ipv6, mtu\",\n        [\tpytest.param(True, False, 1514),\t],\n        )\n    def test_bgp_speaker(localhost, ansible_adhoc, testbed, ipv4, ipv6, mtu):\n        \"\"\"setup bgp speaker on T0 topology and verify routes advertised\n        by bgp speaker is received by T0 TOR\n        \"\"\"\n    \n        hostname = testbed['dut']\n>       ptf_hostname = testbed['ptf']\nE       KeyError: 'ptf'\n\ntest_bgp_speaker.py:40: KeyError\n---------------------------- Captured stdout setup -----------------------------\nSunday 24 November 2019  23:45:21 +0000 (0:00:00.047)       0:00:00.047 ******* \nSunday 24 November 2019  23:45:22 +0000 (0:00:01.041)       0:00:01.088 ******* \nSunday 24 November 2019  23:45:22 +0000 (0:00:00.180)       0:00:01.269 ******* \n--------------------------- Captured stdout teardown ---------------------------\nSunday 24 November 2019  23:45:22 +0000 (0:00:00.271)       0:00:01.540 ******* \nSunday 24 November 2019  23:45:22 +0000 (0:00:00.208)       0:00:01.749 ******* \nSunday 24 November 2019  23:45:22 +0000 (0:00:00.167)       0:00:01.916 ******* \nSunday 24 November 2019  23:45:23 +0000 (0:00:00.129)       0:00:02.046 ******* \nSunday 24 November 2019  23:45:23 +0000 (0:00:00.142)       0:00:02.188 ******* \nSunday 24 November 2019  23:45:23 +0000 (0:00:00.163)       0:00:02.352 ******* \nSunday 24 November 2019  23:45:23 +0000 (0:00:00.130)       0:00:02.483 ******* \n=========================== 1 failed in 3.33 seconds ===========================", "stdout_lines": ["============================= test session starts ==============================", "platform linux2 -- Python 2.7.12, pytest-4.6.5, py-1.8.0, pluggy-0.12.0 -- /usr/bin/python", "cachedir: .pytest_cache", "ansible: 2.0.0.2", "rootdir: /builds2/jenkins2/workspace/SONiC_Test-BGP-Speaker/tests, inifile: pytest.ini", "plugins: ansible-2.0.2", "collecting ... collected 1 item", "", "test_bgp_speaker.py::test_bgp_speaker[True-False-1514] FAILED            [100%]", "", "=================================== FAILURES ===================================", "______________________ test_bgp_speaker[True-False-1514] _______________________", "", "localhost = <pytest_ansible.module_dispatcher.v2.ModuleDispatcherV2 object at 0x7f779dbfae10>", "ansible_adhoc = <function init_host_mgr at 0x7f779dc055f0>", "testbed = {'comment': 'Mellanox MTR testbed', 'dut': 'r-tigris-04', 'group-name': 'vm-t1', 'ptf_image_name': 'docker-ptf-mlnx', ...}", "ipv4 = True, ipv6 = False, mtu = 1514", "", "    @pytest.mark.parametrize(", "        \"ipv4, ipv6, mtu\",", "        [\tpytest.param(True, False, 1514),\t],", "        )", "    def test_bgp_speaker(localhost, ansible_adhoc, testbed, ipv4, ipv6, mtu):", "        \"\"\"setup bgp speaker on T0 topology and verify routes advertised", "        by bgp speaker is received by T0 TOR", "        \"\"\"", "    ", "        hostname = testbed['dut']", ">       ptf_hostname = testbed['ptf']", "E       KeyError: 'ptf'", "", "test_bgp_speaker.py:40: KeyError", "---------------------------- Captured stdout setup -----------------------------", "Sunday 24 November 2019  23:45:21 +0000 (0:00:00.047)       0:00:00.047 ******* ", "Sunday 24 November 2019  23:45:22 +0000 (0:00:01.041)       0:00:01.088 ******* ", "Sunday 24 November 2019  23:45:22 +0000 (0:00:00.180)       0:00:01.269 ******* ", "--------------------------- Captured stdout teardown ---------------------------", "Sunday 24 November 2019  23:45:22 +0000 (0:00:00.271)       0:00:01.540 ******* ", "Sunday 24 November 2019  23:45:22 +0000 (0:00:00.208)       0:00:01.749 ******* ", "Sunday 24 November 2019  23:45:22 +0000 (0:00:00.167)       0:00:01.916 ******* ", "Sunday 24 November 2019  23:45:23 +0000 (0:00:00.129)       0:00:02.046 ******* ", "Sunday 24 November 2019  23:45:23 +0000 (0:00:00.142)       0:00:02.188 ******* ", "Sunday 24 November 2019  23:45:23 +0000 (0:00:00.163)       0:00:02.352 ******* ", "Sunday 24 November 2019  23:45:23 +0000 (0:00:00.130)       0:00:02.483 ******* ", "=========================== 1 failed in 3.33 seconds ==========================="], "warnings": []}
07:45:23 ...ignoring
07:45:23 
07:45:23 TASK [test : debug] ************************************************************
07:45:23 task path: /builds2/jenkins2/workspace/SONiC_Test-BGP-Speaker/ansible/roles/test/tasks/pytest_runner.yml:64
07:45:23 Sunday 24 November 2019  23:45:23 +0000 (0:00:04.452)       0:00:17.849 ******* 
07:45:24 ok: [r-tigris-04-t0] => {
07:45:24     "msg": [
07:45:24         "============================= test session starts ==============================", 
07:45:24         "platform linux2 -- Python 2.7.12, pytest-4.6.5, py-1.8.0, pluggy-0.12.0 -- /usr/bin/python", 
07:45:24         "cachedir: .pytest_cache", 
07:45:24         "ansible: 2.0.0.2", 
07:45:24         "rootdir: /builds2/jenkins2/workspace/SONiC_Test-BGP-Speaker/tests, inifile: pytest.ini", 
07:45:24         "plugins: ansible-2.0.2", 
07:45:24         "collecting ... collected 1 item", 
07:45:24         "", 
07:45:24         "test_bgp_speaker.py::test_bgp_speaker[True-False-1514] FAILED            [100%]", 
07:45:24         "", 
07:45:24         "=================================== FAILURES ===================================", 
07:45:24         "______________________ test_bgp_speaker[True-False-1514] _______________________", 
07:45:24         "", 
07:45:24         "localhost = <pytest_ansible.module_dispatcher.v2.ModuleDispatcherV2 object at 0x7f779dbfae10>", 
07:45:24         "ansible_adhoc = <function init_host_mgr at 0x7f779dc055f0>", 
07:45:24         "testbed = {'comment': 'Mellanox MTR testbed', 'dut': 'r-tigris-04', 'group-name': 'vm-t1', 'ptf_image_name': 'docker-ptf-mlnx', ...}", 
07:45:24         "ipv4 = True, ipv6 = False, mtu = 1514", 
07:45:24         "", 
07:45:24         "    @pytest.mark.parametrize(", 
07:45:24         "        \"ipv4, ipv6, mtu\",", 
07:45:24         "        [\tpytest.param(True, False, 1514),\t],", 
07:45:24         "        )", 
07:45:24         "    def test_bgp_speaker(localhost, ansible_adhoc, testbed, ipv4, ipv6, mtu):", 
07:45:24         "        \"\"\"setup bgp speaker on T0 topology and verify routes advertised", 
07:45:24         "        by bgp speaker is received by T0 TOR", 
07:45:24         "        \"\"\"", 
07:45:24         "    ", 
07:45:24         "        hostname = testbed['dut']", 
07:45:24         ">       ptf_hostname = testbed['ptf']", 
07:45:24         "E       KeyError: 'ptf'", 
07:45:24         "", 
07:45:24         "test_bgp_speaker.py:40: KeyError", 
07:45:24         "---------------------------- Captured stdout setup -----------------------------", 
07:45:24         "Sunday 24 November 2019  23:45:21 +0000 (0:00:00.047)       0:00:00.047 ******* ", 
07:45:24         "Sunday 24 November 2019  23:45:22 +0000 (0:00:01.041)       0:00:01.088 ******* ", 
07:45:24         "Sunday 24 November 2019  23:45:22 +0000 (0:00:00.180)       0:00:01.269 ******* ", 
07:45:24         "--------------------------- Captured stdout teardown ---------------------------", 
07:45:24         "Sunday 24 November 2019  23:45:22 +0000 (0:00:00.271)       0:00:01.540 ******* ", 
07:45:24         "Sunday 24 November 2019  23:45:22 +0000 (0:00:00.208)       0:00:01.749 ******* ", 
07:45:24         "Sunday 24 November 2019  23:45:22 +0000 (0:00:00.167)       0:00:01.916 ******* ", 
07:45:24         "Sunday 24 November 2019  23:45:23 +0000 (0:00:00.129)       0:00:02.046 ******* ", 
07:45:24         "Sunday 24 November 2019  23:45:23 +0000 (0:00:00.142)       0:00:02.188 ******* ", 
07:45:24         "Sunday 24 November 2019  23:45:23 +0000 (0:00:00.163)       0:00:02.352 ******* ", 
07:45:24         "Sunday 24 November 2019  23:45:23 +0000 (0:00:00.130)       0:00:02.483 ******* ", 
07:45:24         "=========================== 1 failed in 3.33 seconds ==========================="
07:45:24     ]
07:45:24 }
```



<!--
Briefly describe the problem you are having in a few paragraphs.
-->

**Steps to reproduce the issue:**
1. Run the BGP speaker testing from ansible playbook
2. The ansible playbook will eventually run the test_bgp_speaker.py using py.test

**Describe the results you received:**
Testing pass.

**Describe the results you expected:**
Testing failed with KeyError: 'ptf'

**Additional information you deem important:**
<!--
software/ASIC/Hardware Flatform version and info
-->
    **Output of `show version`:**

    ```
    (paste your output here)
    ```

    **Attach debug file `sudo generate_dump`:**

    ```
    (paste your output here)
    ```

--- a/tests/test_bgp_speaker.py
+++ b/tests/test_bgp_speaker.py
@@ -1,11 +1,10 @@
 import pytest
 from netaddr import *
-import sys
 import time
-import ipaddress
+import logging
 import requests
-from ansible_host import AnsibleHost
 from ptf_runner import ptf_runner
+
 
 def generate_ips(num, prefix, exclude_ips):
     """
@@ -28,46 +27,48 @@ def generate_ips(num, prefix, exclude_ips):
 
     return generated_ips
 
+
 def announce_route(ptfip, neighbor, route, nexthop, port):
-    url  = "http://%s:%d" % (ptfip, port)
-    data = { "command": "neighbor %s announce route %s next-hop %s" % (neighbor, route, nexthop) }
+    url = "http://%s:%d" % (ptfip, port)
+    data = {"command": "neighbor %s announce route %s next-hop %s" % (neighbor, route, nexthop)}
     r = requests.post(url, data=data)
     assert r.status_code == 200
 
-@pytest.mark.parametrize(
-    "ipv4, ipv6, mtu",
-    [	pytest.param(True, False, 1514),	],
-    )
-def test_bgp(localhost, ansible_adhoc, testbed, ipv4, ipv6, mtu):
+
+@pytest.mark.parametrize("ipv4, ipv6, mtu", [pytest.param(True, False, 1514)])
+def test_bgp(duthost, ptfhost, ipv4, ipv6, mtu):
     """setup bgp speaker on T0 topology and verify routes advertised
     by bgp speaker is received by T0 TOR
     """
 
-    hostname = testbed['dut']
-    ptf_hostname = testbed['ptf']
-    host = AnsibleHost(ansible_adhoc, hostname)
-    ptfhost = AnsibleHost(ansible_adhoc, ptf_hostname)
-    ptfip = ptfhost.host.options['inventory_manager'].get_host(ptf_hostname).vars['ansible_host']
+    ptfip = ptfhost.host.options['inventory_manager'].get_host(ptfhost.hostname).vars['ansible_host']
+    logging.info("ptfip=%s" % ptfip)
 
-    mg_facts = host.minigraph_facts(host=hostname)['ansible_facts']
-    host_facts  = host.setup()['ansible_facts']
+    ptfhost.script("./scripts/remove_ip.sh")
+    ptfhost.script("./scripts/change_mac.sh")
 
-    res = host.shell("sonic-cfggen -m -d -y /etc/sonic/constants.yml -v \"constants.deployment_id_asn_map[DEVICE_METADATA['localhost']['deployment_id']]\"")
+    mg_facts = duthost.minigraph_facts(host=duthost.hostname)['ansible_facts']
+    host_facts = duthost.setup()['ansible_facts']
+
+    res = duthost.shell("sonic-cfggen -m -d -y /etc/sonic/constants.yml -v \"constants.deployment_id_asn_map[DEVICE_METADATA['localhost']['deployment_id']]\"")
     bgp_speaker_asn = res['stdout']
 
-    vlan_ips = generate_ips(3, \
-            "%s/%s" % (mg_facts['minigraph_vlan_interfaces'][0]['addr'], mg_facts['minigraph_vlan_interfaces'][0]['prefixlen']),
-            [IPAddress(mg_facts['minigraph_vlan_interfaces'][0]['addr'])])
+    vlan_ips = generate_ips(3,
+                            "%s/%s" % (mg_facts['minigraph_vlan_interfaces'][0]['addr'],
+                                       mg_facts['minigraph_vlan_interfaces'][0]['prefixlen']),
+                            [IPAddress(mg_facts['minigraph_vlan_interfaces'][0]['addr'])])
+    logging.info("Generated vlan_ips: %s" % str(vlan_ips))
 
     # three speaker ips, two from peer range, another is vlan ip [0]
     speaker_ips = generate_ips(2, mg_facts['minigraph_bgp_peers_with_range'][0]['ip_range'][0], [])
     speaker_ips.append(vlan_ips[0])
+    logging.info("speaker_ips: %s" % str(speaker_ips))
 
     for ip in vlan_ips:
-        host.command("ip route flush %s/32" % ip.ip)
-        host.command("ip route add %s/32 dev %s" % (ip.ip, mg_facts['minigraph_vlan_interfaces'][0]['attachto']))
+        duthost.command("ip route flush %s/32" % ip.ip)
+        duthost.command("ip route add %s/32 dev %s" % (ip.ip, mg_facts['minigraph_vlan_interfaces'][0]['attachto']))
 
-    root_dir   = "/root"
+    root_dir = "/root"
     port_num = [5000, 6000, 7000]
 
     lo_addr = mg_facts['minigraph_lo_interfaces'][0]['addr']
@@ -78,8 +79,9 @@ def test_bgp(localhost, ansible_adhoc, testbed, ipv4, ipv6, mtu):
     vlan_ports = []
     for i in range(0, 3):
         vlan_ports.append(mg_facts['minigraph_port_indices'][mg_facts['minigraph_vlans'][mg_facts['minigraph_vlan_interfaces'][0]['attachto']]['members'][i]])
+    logging.info("vlan_ports: %s" % str(vlan_ports))
 
-    # setup ip/routes in ptf
+    logging.info("setup ip/routes in ptf")
     ptfhost.shell("ifconfig eth%d %s" % (vlan_ports[0], vlan_ips[0]))
     ptfhost.shell("ifconfig eth%d:0 %s" % (vlan_ports[0], speaker_ips[0]))
     ptfhost.shell("ifconfig eth%d:1 %s" % (vlan_ports[0], speaker_ips[1]))
@@ -93,8 +95,8 @@ def test_bgp(localhost, ansible_adhoc, testbed, ipv4, ipv6, mtu):
     lo_addr = mg_facts['minigraph_lo_interfaces'][0]['addr']
     for i in range(0, 3):
         local_ip = str(speaker_ips[i].ip)
-        ptfhost.exabgp(name="bgps%d" % i, \
-                       state="started", \
+        ptfhost.exabgp(name="bgps%d" % i,
+                       state="started",
                        local_ip=local_ip,
                        router_id=local_ip,
                        peer_ip=lo_addr,
@@ -107,55 +109,69 @@ def test_bgp(localhost, ansible_adhoc, testbed, ipv4, ipv6, mtu):
     peer_range = mg_facts['minigraph_bgp_peers_with_range'][0]['ip_range'][0]
     prefix = '10.10.10.0/26'
 
-    # announce route
+    logging.info("announce route")
     announce_route(ptfip, lo_addr, prefix, vlan_ips[1].ip, port_num[0])
     announce_route(ptfip, lo_addr, prefix, vlan_ips[2].ip, port_num[1])
     announce_route(ptfip, lo_addr, peer_range, vlan_ips[0].ip, port_num[2])
 
-    # make sure routes announced to dynamic bgp neighbors
+    logging.info("Wait some time to make sure routes announced to dynamic bgp neighbors")
     time.sleep(60)
 
-    bgp_facts = host.bgp_facts()['ansible_facts']
+    # The ping here is workaround for known issue:
+    #     https://github.com/Azure/SONiC/issues/387 Pre-ARP support for static route config
+    # When there is no arp entry for next hop, routes learnt from exabgp will not be set down to ASIC
+    # Traffic to prefix 10.10.10.0 will be routed to vEOS VMs via default gateway.
+    duthost.shell("ping %s -c 3" % vlan_ips[1].ip)
+    duthost.shell("ping %s -c 3" % vlan_ips[2].ip)
+    time.sleep(5)
 
-    # Verify bgp sessions are established
+    bgp_facts = duthost.bgp_facts()['ansible_facts']
+
+    logging.info("Verify bgp sessions are established")
     for k, v in bgp_facts['bgp_neighbors'].items():
         assert v['state'] == 'established'
 
-    # Verify accepted prefixes of the dynamic neighbors are correct
+    logging.info("Verify accepted prefixes of the dynamic neighbors are correct")
     for ip in speaker_ips:
         assert bgp_facts['bgp_neighbors'][str(ip.ip)]['accepted prefixes'] == 1
     assert bgp_facts['bgp_neighbors'][str(vlan_ips[0].ip)]['accepted prefixes'] == 1
 
-    ## Run ptf test
-    # Generate route-port map information
-    extra_vars = \
-                { 'announce_prefix': '10.10.10.0/26',
-                  'minigraph_portchannels'  : mg_facts['minigraph_portchannels'],
-                  'minigraph_vlans'  : mg_facts['minigraph_vlans'],
-                  'minigraph_port_indices'  : mg_facts['minigraph_port_indices']}
+    logging.info("Generate route-port map information")
+    extra_vars = {'announce_prefix': '10.10.10.0/26',
+                  'minigraph_portchannels': mg_facts['minigraph_portchannels'],
+                  'minigraph_vlans': mg_facts['minigraph_vlans'],
+                  'minigraph_port_indices': mg_facts['minigraph_port_indices']}
     ptfhost.host.options['variable_manager'].extra_vars.update(extra_vars)
+    logging.info("extra_vars: %s" % str(ptfhost.host.options['variable_manager'].extra_vars))
 
     ptfhost.template(src="bgp_speaker/bgp_speaker_route.j2", dest="/root/bgp_speaker_route.txt")
 
     ptfhost.copy(src="ptftests", dest=root_dir)
 
-    ptf_runner(ptfhost, \
-               "ptftests",
-               "fib_test.FibTest",
-               platform_dir="ptftests",
-               params={"testbed_type": "t0",
-                      "router_mac": host_facts['ansible_Ethernet0']['macaddress'],
-                      "fib_info": "/root/bgp_speaker_route.txt",
-                      "ipv4": ipv4,
-                      "ipv6": ipv6,
-                      "testbed_mtu": mtu },
-               log_file="/tmp/bgp_speaker_test.FibTest.log",
-               socket_recv_size=16384)
+    logging.info("run ptf test")
+    try:
+        ptf_runner(ptfhost,
+                   "ptftests",
+                   "fib_test.FibTest",
+                   platform_dir="ptftests",
+                   params={"testbed_type": "t0",
+                           "router_mac": host_facts['ansible_Ethernet0']['macaddress'],
+                           "fib_info": "/root/bgp_speaker_route.txt",
+                           "ipv4": ipv4,
+                           "ipv6": ipv6,
+                           "testbed_mtu": mtu },
+                   log_file="/tmp/bgp_speaker_test.FibTest.log",
+                   socket_recv_size=16384)
+    except Exception as e:
+        logging.error("Run ptf test exception: %s" % repr(e))
+        assert False, "PTF test for bgp_speaker failed"
 
-    for i in range(0, 3):
-        ptfhost.exabgp(name="bgps%d" % i, state="absent")
+    finally:
+        logging.info("cleanup")
+        for i in range(0, 3):
+            ptfhost.exabgp(name="bgps%d" % i, state="absent")
 
-    for ip in vlan_ips:
-        host.command("ip route flush %s/32" % ip.ip)
+        for ip in vlan_ips:
+            duthost.command("ip route flush %s/32" % ip.ip)
 
-    ptfhost.shell("ip addr flush dev eth{}".format(mg_facts['minigraph_port_indices'][mg_facts['minigraph_vlans'][mg_facts['minigraph_vlan_interfaces'][0]['attachto']]['members'][0]]))
+        ptfhost.shell("ip addr flush dev eth{}".format(mg_facts['minigraph_port_indices'][mg_facts['minigraph_vlans'][mg_facts['minigraph_vlan_interfaces'][0]['attachto']]['members'][0]]))

--- a/tests/test_bgp_speaker.py
+++ b/tests/test_bgp_speaker.py
@@ -161,7 +161,6 @@ def test_bgp_speaker_announce_routes(common_setup_teardown, duthost, ptfhost, ip
     bgp_facts = duthost.bgp_facts()['ansible_facts']
     for ip in speaker_ips:
         assert bgp_facts['bgp_neighbors'][str(ip.ip)]['accepted prefixes'] == 1
-    assert bgp_facts['bgp_neighbors'][str(vlan_ips[0].ip)]['accepted prefixes'] == 1
 
     logging.info("Generate route-port map information")
     extra_vars = {'announce_prefix': '10.10.10.0/26',

--- a/tests/test_bgp_speaker.py
+++ b/tests/test_bgp_speaker.py
@@ -35,11 +35,10 @@ def announce_route(ptfip, neighbor, route, nexthop, port):
     assert r.status_code == 200
 
 
-@pytest.mark.parametrize("ipv4, ipv6, mtu", [pytest.param(True, False, 1514)])
-def test_bgp(duthost, ptfhost, ipv4, ipv6, mtu):
-    """setup bgp speaker on T0 topology and verify routes advertised
-    by bgp speaker is received by T0 TOR
-    """
+@pytest.fixture(scope="module")
+def common_setup_teardown(duthost, ptfhost):
+
+    logging.info("########### Setup for bgp speaker testing ###########")
 
     ptfip = ptfhost.host.options['inventory_manager'].get_host(ptfhost.hostname).vars['ansible_host']
     logging.info("ptfip=%s" % ptfip)
@@ -48,18 +47,16 @@ def test_bgp(duthost, ptfhost, ipv4, ipv6, mtu):
     ptfhost.script("./scripts/change_mac.sh")
 
     mg_facts = duthost.minigraph_facts(host=duthost.hostname)['ansible_facts']
-    host_facts = duthost.setup()['ansible_facts']
+    interface_facts = duthost.interface_facts()['ansible_facts']
 
     res = duthost.shell("sonic-cfggen -m -d -y /etc/sonic/constants.yml -v \"constants.deployment_id_asn_map[DEVICE_METADATA['localhost']['deployment_id']]\"")
     bgp_speaker_asn = res['stdout']
 
-    vlan_ips = generate_ips(3,
-                            "%s/%s" % (mg_facts['minigraph_vlan_interfaces'][0]['addr'],
-                                       mg_facts['minigraph_vlan_interfaces'][0]['prefixlen']),
+    vlan_ips = generate_ips(3, "%s/%s" % (mg_facts['minigraph_vlan_interfaces'][0]['addr'],
+                                          mg_facts['minigraph_vlan_interfaces'][0]['prefixlen']),
                             [IPAddress(mg_facts['minigraph_vlan_interfaces'][0]['addr'])])
     logging.info("Generated vlan_ips: %s" % str(vlan_ips))
 
-    # three speaker ips, two from peer range, another is vlan ip [0]
     speaker_ips = generate_ips(2, mg_facts['minigraph_bgp_peers_with_range'][0]['ip_range'][0], [])
     speaker_ips.append(vlan_ips[0])
     logging.info("speaker_ips: %s" % str(speaker_ips))
@@ -68,7 +65,6 @@ def test_bgp(duthost, ptfhost, ipv4, ipv6, mtu):
         duthost.command("ip route flush %s/32" % ip.ip)
         duthost.command("ip route add %s/32 dev %s" % (ip.ip, mg_facts['minigraph_vlan_interfaces'][0]['attachto']))
 
-    root_dir = "/root"
     port_num = [5000, 6000, 7000]
 
     lo_addr = mg_facts['minigraph_lo_interfaces'][0]['addr']
@@ -92,7 +88,7 @@ def test_bgp(duthost, ptfhost, ipv4, ipv6, mtu):
     ptfhost.shell("ip route flush %s/%d" % (lo_addr, lo_addr_prefixlen))
     ptfhost.shell("ip route add %s/%d via %s" % (lo_addr, lo_addr_prefixlen, vlan_addr))
 
-    lo_addr = mg_facts['minigraph_lo_interfaces'][0]['addr']
+    logging.info("Start exabgp on ptf")
     for i in range(0, 3):
         local_ip = str(speaker_ips[i].ip)
         ptfhost.exabgp(name="bgps%d" % i,
@@ -104,18 +100,54 @@ def test_bgp(duthost, ptfhost, ipv4, ipv6, mtu):
                        peer_asn=mg_facts['minigraph_bgp_asn'],
                        port=str(port_num[i]))
 
-    time.sleep(10)
+    logging.info("########### Done setup for bgp speaker testing ###########")
 
-    peer_range = mg_facts['minigraph_bgp_peers_with_range'][0]['ip_range'][0]
-    prefix = '10.10.10.0/26'
+    yield ptfip, mg_facts, interface_facts, vlan_ips, speaker_ips, port_num
+
+    logging.info("########### Teardown for bgp speaker testing ###########")
+
+    for i in range(0, 3):
+        ptfhost.exabgp(name="bgps%d" % i, state="absent")
+
+    for ip in vlan_ips:
+        duthost.command("ip route flush %s/32" % ip.ip, module_ignore_errors=True)
+
+    ptfhost.script("./scripts/remove_ip.sh")
+
+    logging.info("########### Done teardown for bgp speaker testing ###########")
+
+
+def test_bgp_speaker_bgp_sessions(common_setup_teardown, duthost, ptfhost):
+    """Setup bgp speaker on T0 topology and verify bgp sessions are established
+    """
+    ptfip, mg_facts, interface_facts, vlan_ips, speaker_ips, port_num = common_setup_teardown
+
+    logging.info("Wait some time to verify that bgp sessions are established")
+    time.sleep(20)
+    bgp_facts = duthost.bgp_facts()['ansible_facts']
+    assert all([v["state"] == "established" for _, v in bgp_facts["bgp_neighbors"].items()]), \
+        "Not all bgp sessions are established"
+    assert str(speaker_ips[2].ip) in bgp_facts["bgp_neighbors"], "No bgp session with PTF"
+
+
+@pytest.mark.parametrize("ipv4, ipv6, mtu", [pytest.param(True, False, 1514)])
+def test_bgp_speaker_announce_routes(common_setup_teardown, duthost, ptfhost, ipv4, ipv6, mtu):
+    """Setup bgp speaker on T0 topology and verify routes advertised by bgp speaker is received by T0 TOR
+
+    """
+    ptfip, mg_facts, interface_facts, vlan_ips, speaker_ips, port_num = common_setup_teardown
 
     logging.info("announce route")
+    peer_range = mg_facts['minigraph_bgp_peers_with_range'][0]['ip_range'][0]
+    lo_addr = mg_facts['minigraph_lo_interfaces'][0]['addr']
+    lo_addr_prefixlen = int(mg_facts['minigraph_lo_interfaces'][0]['prefixlen'])
+    prefix = '10.10.10.0/26'
     announce_route(ptfip, lo_addr, prefix, vlan_ips[1].ip, port_num[0])
     announce_route(ptfip, lo_addr, prefix, vlan_ips[2].ip, port_num[1])
     announce_route(ptfip, lo_addr, peer_range, vlan_ips[0].ip, port_num[2])
 
     logging.info("Wait some time to make sure routes announced to dynamic bgp neighbors")
-    time.sleep(60)
+    time.sleep(30)
 
     # The ping here is workaround for known issue:
     #     https://github.com/Azure/SONiC/issues/387 Pre-ARP support for static route config
@@ -125,13 +157,8 @@ def test_bgp(duthost, ptfhost, ipv4, ipv6, mtu):
     duthost.shell("ping %s -c 3" % vlan_ips[2].ip)
     time.sleep(5)
 
-    bgp_facts = duthost.bgp_facts()['ansible_facts']
-
-    logging.info("Verify bgp sessions are established")
-    for k, v in bgp_facts['bgp_neighbors'].items():
-        assert v['state'] == 'established'
-
     logging.info("Verify accepted prefixes of the dynamic neighbors are correct")
+    bgp_facts = duthost.bgp_facts()['ansible_facts']
     for ip in speaker_ips:
         assert bgp_facts['bgp_neighbors'][str(ip.ip)]['accepted prefixes'] == 1
     assert bgp_facts['bgp_neighbors'][str(vlan_ips[0].ip)]['accepted prefixes'] == 1
@@ -146,32 +173,19 @@ def test_bgp(duthost, ptfhost, ipv4, ipv6, mtu):
 
     ptfhost.template(src="bgp_speaker/bgp_speaker_route.j2", dest="/root/bgp_speaker_route.txt")
 
-    ptfhost.copy(src="ptftests", dest=root_dir)
+    ptfhost.copy(src="ptftests", dest="/root")
 
     logging.info("run ptf test")
-    try:
-        ptf_runner(ptfhost,
-                   "ptftests",
-                   "fib_test.FibTest",
-                   platform_dir="ptftests",
-                   params={"testbed_type": "t0",
-                           "router_mac": host_facts['ansible_Ethernet0']['macaddress'],
-                           "fib_info": "/root/bgp_speaker_route.txt",
-                           "ipv4": ipv4,
-                           "ipv6": ipv6,
-                           "testbed_mtu": mtu },
-                   log_file="/tmp/bgp_speaker_test.FibTest.log",
-                   socket_recv_size=16384)
-    except Exception as e:
-        logging.error("Run ptf test exception: %s" % repr(e))
-        assert False, "PTF test for bgp_speaker failed"
 
-    finally:
-        logging.info("cleanup")
-        for i in range(0, 3):
-            ptfhost.exabgp(name="bgps%d" % i, state="absent")
-
-        for ip in vlan_ips:
-            duthost.command("ip route flush %s/32" % ip.ip)
-
-        ptfhost.shell("ip addr flush dev eth{}".format(mg_facts['minigraph_port_indices'][mg_facts['minigraph_vlans'][mg_facts['minigraph_vlan_interfaces'][0]['attachto']]['members'][0]]))
+    ptf_runner(ptfhost,
+                "ptftests",
+                "fib_test.FibTest",
+                platform_dir="ptftests",
+                params={"testbed_type": "t0",
+                        "router_mac": interface_facts['ansible_interface_facts']['Ethernet0']['macaddress'],
+                        "fib_info": "/root/bgp_speaker_route.txt",
+                        "ipv4": ipv4,
+                        "ipv6": ipv6,
+                        "testbed_mtu": mtu },
+                log_file="/tmp/bgp_speaker_test.FibTest.log",
+                socket_recv_size=16384)

